edit_course/operations/configure.py

```diff
@@ -356,299 +357,357 @@ def get_build_log(instance):
+                dt = parse_date(m["late_close"], errors)
+                if dt:
+                    module.late_submission_deadline = dt
+                    module.late_submissions_allowed = True
+            elif "late_duration" in m:
+                dt = parse_duration(module.closing_time, m["late_duration"], errors)
+                if dt:
+                    module.late_submission_deadline = dt
+                    module.late_submissions_allowed = True
+            if "late_penalty" in m:
+                f = parse_float(m["late_penalty"], errors)
+                if not f is None:
+                    module.late_submission_penalty = f
+
+            module.full_clean()
+            module.save()
+            seen_modules.append(module.id)
+
+            if not ("numerate_ignoring_modules" in config \
+                    and parse_bool(config["numerate_ignoring_modules"])):
+                nn = 0
+            if "children" in m:
+                nn = configure_learning_objects(category_map, module, m["children"],
+                    None, seen_objects, errors, nn)
+
+        for module in instance.course_modules.all():
+            # cache invalidation uses the parent when learning object is saved:
+            # prefetch parent so that it wont be fetched after the it was deleted
+            for lobject in module.learning_objects.all().prefetch_related('parent'):
+                if lobject.id not in seen_objects:
+                    exercise = lobject.as_leaf_class()
+                    if (
+                        not isinstance(exercise, BaseExercise)
+                        or exercise.submissions.count() == 0
+                    ):
+                        exercise.delete()
+                    else:
+                        lobject.status = LearningObject.STATUS.HIDDEN
+                        lobject.order = 9999
+                        # .parent may have been deleted: only save status and order
+                        lobject.save(update_fields=["status", "order"])
+            if module.id not in seen_modules:
+                if module.learning_objects.count() == 0:
+                    module.delete()
                 else:
-                    lobject.status = LearningObject.STATUS.HIDDEN
-                    lobject.order = 9999
-                    lobject.save()
-        if not module.id in seen_modules:
-            if module.learning_objects.count() == 0:
-                module.delete()
+                    module.status = CourseModule.STATUS.HIDDEN
+                    module.save()
+
+        # Clean up obsolete categories.
+        for category in instance.categories.filter(status=LearningObjectCategory.STATUS.HIDDEN):
+            if category.learning_objects.count() == 0:
+                category.delete()
+
+        if "publish_url" in config:
+            success = False
+            publish_errors = []
+            try:
+                response = aplus_auth.get(config["publish_url"])
```

Comment:
Where is this HTTP request typically going to? Git manager? And why is this needed?

Reply:
Like I responded to another comment: It is used to notify gitmanager (or whatever service we are fetching the config from) that the configuration seems OK and now we should publish the changes. The point is that gitmanager does not publish (roll out) the changes before we know that A+ is ok with them, and it also means that A+ does its preprocessing _before_ any changes are published, so downtime (or time when the course is invalid, in some sense) is short.